js/models/latlng.ts

```diff
@@ -9,3 +9,28 @@ export const LatLng = z.object({
   latitude: z.number(),
   longitude: z.number(),
 }) satisfies z.ZodType<LatLng>;
+
+/**
+ * Find how far between a start an finish and intermediate point is,
+ * by projecting it onto the vector between the start and finish.
+ *
+ * @param start A coordinate pair
+ * @param finish Another coordinate pair
+ * @param point A target pair
+ * @returns How far between start and finish `point` is, between 0 and 1.
+ */
+export const proportionBetweenLatLngs = (
+  start: LatLng,
+  finish: LatLng,
+  point: LatLng,
+): number => {
+  const totalLatDelta: number = finish.latitude - start.latitude;
+  const totalLngDelta: number = finish.longitude - start.longitude;
+  const pointLatDelta: number = point.latitude - start.latitude;
+  const pointLngDelta: number = point.longitude - start.longitude;
+  const dotProductStartToPoint: number =
+    totalLatDelta * pointLatDelta + totalLngDelta * pointLngDelta;
+  const startToFinishMagnitudeSquared: number =
+    totalLatDelta * totalLatDelta + totalLngDelta * totalLngDelta;
+  return dotProductStartToPoint / startToFinishMagnitudeSquared;
```

Comment:
Do we want to guard against possible divide by zero here? In the event that `start` and `finish` are the same (arguably a misuse), this risks spitting out `NaN`. Maybe worth checking the denominator and returning `0` in that case.

Reply:
(Whoops - realizing this is in `main` already, so not part of this PR. Question still stands?)